### ISSUE TITLE
Build health metrics dimensions

### ISSUE THREAD

**Issue by mrh666** [jenkinsci/opentelemetry-plugin#952]
### What feature do you want to see added?

@cyrille-leclerc on that page https://plugins.jenkins.io/opentelemetry/ there is a screenshot of kibana https://raw.githubusercontent.com/jenkinsci/opentelemetry-plugin/master/docs/images/kibana_jenkins_overview_dashboard.png with all graphs I need, e.g. job duration, failed steps, long steps, etc. How can we get those metrics exported to Dynatrace?

**Comment by cyrille-leclerc** [comment]
A solution could be to produce a metric like
```
ci.pipeline.run.duration{
   ci.pipeline.id="/my-team/my-war/master", //pipeline full name
   // see hudson.model.Run#getResult() 
   // SUCCESS, UNSTABLE, FAILURE, NOT_BUILT, ABORTED
   ci.pipeline.run.result="SUCCESS"

**Comment by mrh666** [comment]
`Could it be possible with allo/deny lists based on regex to just collect metrics on the pipelines that matter to you?`

This one is really important!

**Comment by christophe-kamphaus-jemmic** [comment]
> I was waiting for the OTel CI/CD SIG to standardize such a metric

Indeed work in the OTel CI/CD SIG related to metrics is in progress.
We are currently standardizing metrics related to VCS and we plan to follow that up with metrics related to pipelines, queues and agents.
`ci.pipeline.run.duration: histogram` sounds like a good metric. I will propose it in the SIG.

One issue I can see with using a histogram is that the chosen buckets might not give enough insight to take any action. Some jobs might be of very short duration, while others could take hours or even days to complete.
So how would you define the buckets?

I had very good success in using metrics queries on traces/spans for job duration as well as stage duration using the steps I introduced in https://github.com/jenkinsci/opentelemetry-plugin/pull/827 (example use in a pipeline here: https://github.com/jenkinsci/opentelemetry-plugin/issues/811#issuecomment-2116113648). This allowed me to have very detailed statistics (eg. average duration per day or job) and is filterable per job.

> I'm thinking of helping Jenkins admins control cardinality of such metrics enabling allow & deny lists of pipeline names as we have seen Jenkins instances with thousands of pipelines.

For sure cardinality is an issue when the number of time series scales with a dynamic value like the number of jobs managed by Jenkins. It's not as bad as when we would have a separate time series per build, but still it needs to be managed. ([prometheus-plugin](https://github.com/jenkinsci/prometheus-plugin/blob/master/docs/metrics/index.md) has per-build metrics guarded by a checkbox config option)

Controlling which jobs generate this metric on Jenkins-side I think is a very good option.

Alternatively it's also possible to filter later:
* using metric relabelling rules in Prometheus or ServiceMonitor
* with opentelemetry-collector a [filterprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md) could be used
  ```
  processors:
    filter/ci:
      error_mode: ignore
      metrics:
        metric:
            - 'name == "ci.pipeline.run.duration" and not(IsMatch(attributes["ci.pipeline.id"], "my-otel-pipelines.*"))'
  ```

**Comment by cyrille-leclerc** [comment]
Cc @miraccan00

**Comment by cyrille-leclerc** [comment]
Please use the `ci.pipeline.run.duration{ci.pipeline.id="<<pipeline full name>>", ci.pipeline.result="<<SUCCESS, UNSTABLE, FAILURE, NOT_BUILT, ABORTED>>"}` histogram metric we have just released.
ℹ Use the `otel.instrumentation.jenkins.run.metric.duration.allow_list` and `otel.instrumentation.jenkins.run.metric.duration.deny_list` to specify the pipelines for which you want to capture the run duration, other pipelines will be aggregated in the `ci.pipeline.id="#other#"` time series.

See documentation https://github.com/jenkinsci/opentelemetry-plugin/blob/main/docs/monitoring-metrics.md#build-duration

I'm marking your enhancement request as solved. Please open new enhancement requests if needed.

**Comment by cyrille-leclerc** [comment]
Thanks for testing @mrh666 .
Cn you please provide a list of example job full display names, including the parent folders + the filter you want to apply so we can test?

**Comment by mrh666** [comment]
For example, the simple job: https://xxxxxxx/job/telemetry%20test%20pipe/
And I want to have metrics from this job only. Something like '^.*telemetry.*$' in allow_list

**Comment by cyrille-leclerc** [comment]
@mrh666 please just configure:
```
otel.instrumentation.jenkins.run.metric.duration.allow_list=.*telemetry.*
```

Don't put `^` or `$`.
I successfully tested with a pipeline called `/telemetry/test pipe`

**Comment by mrh666** [comment]
@cyrille-leclerc Maybe I'm doing something wrong? Here is settings:
![Image](https://github.com/user-attachments/assets/0d967ad0-6d74-4087-90c6-dbee7c9785ff)

Here is all metrics coming to Dynatrace perfectly except ci.pipeline.run.duration
![Image](https://github.com/user-attachments/assets/eb9795ed-3211-4eda-b0de-e621a4a92c8e)


Versions:
Jenkins 2.462.1
OpenTelemetry API Plugin 1.43.0-38.v1a_9b_53e3f70f
OpenTelemetry Plugin 3.1423.v0d1a_2fcd2429

**Comment by cyrille-leclerc** [comment]
Can you verify that you have a histogram metric `ci.pipeline.duration`, at least reporting the attribute `ci.pipeline.id=#other#` ? I cannot see this in your screenshot.
IF you don't se this metric, then the problem is broader than the `otel.instrumentation.jenkins.run.metric.duration.allow_list` parameter.

**Comment by mrh666** [comment]
@cyrille-leclerc 
Nope, I can't see the metric ci.pipeline.duration in logs or in target Dynatrace. 
Yes, I can see the attribute you mentioned in trace:
```
Dec 09, 2024 10:37:42 AM FINE io.jenkins.plugins.opentelemetry.job.action.AbstractMonitoringAction

Purge span='BUILD telemetry test pipe', spanId=acdda7a6b8db2192, traceId=232f13dd225862520dc9383406c37be1: SpanAndScopes{span=SdkSpan{traceId=232f13dd225862520dc9383406c37be1, spanId=acdda7a6b8db2192, parentSpanContext=ImmutableSpanContext{traceId=232f13dd225862520dc9383406c37be1, spanId=51eff770ee371c5f, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=true, valid=true}, name=BUILD telemetry test pipe, kind=SERVER, attributes=AttributesMap{data={type=job, ci.pipeline.run.completed=true, ci.pipeline.name=telemetry test pipe, ci.pipeline.run.committers=[], ci.pipeline.run.cause=[UserIdCause:my@mail], ci.pipeline.run.url=https://xxxxx/job/telemetry%20test%20pipe/197/, ci.pipeline.type=workflow, ci.pipeline.run.result=SUCCESS, ci.pipeline.run.number=197, ci.pipeline.run.durationMillis=1674, ci.pipeline.id=telemetry test pipe}, capacity=128, totalAddedValues=11}, status=ImmutableStatusData{statusCode=OK, description=SUCCESS}, totalRecordedEvents=0, totalRecordedLinks=0, startEpochNanos=1733740660699406857, endEpochNanos=1733740662393067097}, scopes=0, scopeStartThreadName='Executor #-1 for Built-In Node : executing telemetry test pipe #197'}
```

**Comment by cyrille-leclerc** [comment]
@mrh666 it seems that Dynatrace just introduced support for OTel histogram metrics:
https://www.dynatrace.com/news/blog/opentelemetry-histograms-reveal-patterns-outliers-and-trends/
Could your version precede this improvement?

**Comment by cyrille-leclerc** [comment]
Friendly ping @mrh666
